### ISSUE TITLE
Emulate Memory Card - does not work without SDcard in SD2SP2

### ISSUE THREAD

**Issue by BorgNumberOne** [emukidid/swiss-gc#611]
Hi,

today I would like to run my Swiss-ed (v0.5r1196) GameCube without a SDcard in SD2SP2 anymore - just with GCLoader and a 2GB (non-SDHC) SDcard in Slot-A and a memory card in Slot-B.

**So my current/today GameCube configuration is:**

- GCLoader (2.0.0 beta) with a 400GByte micro SDcard (microSD-to-normal-SD-adapter), FAT32 formatted, all my backed up .iso games are stored  on it
- Slot A: 2GB SD (non-HC/non-XC) FAT32 formated (homebrew and experimental storage)
- just the SD2SP2 adapter left plugged in the GameCube, but without a SDcard anymore
- Slot B: memory card
- Swiss.ini:
```
# Swiss Configuration File!
# Anything written in here will be lost!

#!!Swiss Settings Start!!
SD/IDE Speed=32MHz
Swiss Video Mode=Auto
Enable Debug=No
Hide Unknown file types=No
Stop DVD Motor on startup=No
Enable WiiRD debug=No
Enable File Management=Yes
Disable Video Patches=None
Force Video Active=No
Force DTV Status=No
SMBUserName=
SMBPassword=
SMBShareName=
SMBHostIP=
AutoCheats=No
InitNetwork=No
IGRType=Disabled
AVECompat=GCVideo
FileBrowserType=Standard
BS2Boot=Yes
FTPUserName=
FTPPassword=
FTPHostIP=
FTPPort=0
FTPUsePasv=No
FSPHostIP=
FSPPort=0
FSPPassword=
ShowHiddenFiles=Yes
RecentListLevel=On
GCLoaderTopVersion=2.0.0.BETA
Autoload=
Recent_0=gcldr://Gauntlet - Dark Legacy (Europe) (En,Fr,De).iso
Recent_1=gcldr://GCCP01-FINAL FANTASY Crystal Chronicles.iso
Recent_2=
Recent_3=
Recent_4=
Recent_5=
Recent_6=
Recent_7=
Force Video Mode=No
Force Horizontal Scale=Auto
Force Vertical Offset=+0
Force Vertical Filter=Auto
Disable Alpha Dithering=No
Force Anisotropic Filter=No
Force Widescreen=No
Invert Camera Stick=No
Emulate Read Speed=No
Emulate Memory Card=Yes
#!!Swiss Settings End!!
```

To have a clean Swiss environment (and for reproducing the bug):
- I deleted every swiss.ini and the hidden "swiss_patches" folder(s) on all SDcards before/again.
- the "Configuration Device" (settings) / "Config Medium" is set to: SD card Slot-A.
(So, Swiss settings, game patches and emulated memory card should be stored here on the SD card in Slot-A **without any context to a SDcard in SD2SP2 anymore - logically, normally**)

To see if the emulated memory card is working
(and to get all save games from real memory card in Slot-B to the emulated Memory Card), 
you should normally start any game with:
- following swiss settings: "Emulate Memory Card" and: "Boot through IPL:" switched on
- holding the A button while starting a game and showing GameCube animation --> to get into GameCube's IPL and its memory card management

Now, **Memory Card A** in the GameCube's Memory Card management **is the emulated Memory Card** and you can copy the save games from the Real Memory Card B to **the emulated one (A)**.

These steps did work until yesterday - when a SD card was plugged in to the SD2SP2 adapter and Swiss' Configuration Device was set to: "SD2SP2" or to: "SD Card A" (while a SD card was plugged into the SD2SP2 slot)
So, Swiss v0.5r1196 did work very well (while additionally using a SDcard in SD2SP2)

Today, I removed the SD card from SD2SP2 (prepared a clean Swiss environment - as described above) and I tried many different games (e.g.: Gauntlet or Mario Kart or anything else) from GCLoader with the aim of not using a SD card in SD2SP2 anymore. 

But suddenly, games accidentally/coincidentally/sometimes do not work / boot anymore and the **[SDCard A]\swiss_patches\MemoryCardX.xxx.raw** has a weird file size (sometimes 4,xxMbyte, sometimes 3,xxMbytes, sometimes 8,xx Mbytes, etc.)

If I just put in an empty SDcard into the SD2SP2 adapter and start any game again, everything works fine.
It seems to be that Swiss still has a kind of a context to a SDcard in a SD2SP2 adapter even if the "Configuration Device" **is not set** to SD2SP2 anymore.

P.S. - improvement:
"Configuration Device" (settings) / "Config Medium" (swiss device info dialog) --> two descriptions for one thing? ;)

**Comment by Extrems** [comment]
The patch device and configuration device are two separate things, with the former not being user-selectable and isn't saved anywhere.

It's currently not possible to emulate a memory card in slot A with an SD card in slot A. The SD card must be in slot B.

**Comment by BorgNumberOne** [comment]
There is a strange thing:
Some minutes ago, I tried any game (e.g. Gauntled) without SD2SP2 and Swiss creates a **[SDcard A]\swiss_patches\MemoryCardX.xxx.raw** folder/file structure.

But why?

As you wrote: "It's currently not possible to emulate a memory card in slot A with an SD card in slot A.", why does swiss creates a **[SDcard A]\swiss_patches\MemoryCardX.xxx.raw** folder/file structure.** then?

Is this behavior (checking for SD2SP2 and then SDcard Slot A, Slot B and then firstly using SD2SP2 instead SDcard Slot A for swiss_patches file/folder structure) a completely automatic process independent from Swiss.ini settings?

In short words: if I would like to manage savegames between a real and Swiss' emulated memory card there is a SDcard in SD2SP2 necessary. Is not it?

**Comment by Extrems** [comment]
It's completely automatic, and MemoryCardA isn't supposed to be created in this case.

If the device the game is booted from can function as a patch device (i.e. SD Card Adapters, IDE-EXI), then it necessarily is the patch device.

For other devices (i.e. GC Loader), it picks Serial Port 2, then Memory Slot B, then Memory Slot A.

**Comment by BorgNumberOne** [comment]
Would it be possible in future to realize a MemoryCardA/B emulation and swiss_patches file/folder structure also on GCloader's used/mounted SDcard (Firmware 2.0.0 beta)

Without opening a new issue, why is the transfer speed so different fast/slow
a: copy files from GCloader to a SD2SP2 or SDCardA/B device
b: copy files from a SD2SP2 or SDCardA/B device to GCLoader

**Comment by Extrems** [comment]
The GC Loader write support is unfortunately very slow and hard on your SD card's write endurance. I'm hoping the firmware will improve in the future, but this first implementation was the path of least resistance, and it took forever to even get there.

I'm hoping to begin work on the hypervisor bits for the GC Loader 2.0.0 soon, but SD Card Adapters will remain preferred as patch device.
